### ISSUE TITLE
Automatically handle `KnownTypeAttribute`

### ISSUE THREAD

**Issue by onionhammer** [riok/mapperly#330]
[KnownType] attribute is used on a base type to inform callers/serializers etc what its known derived types are, Mapperly could use this to automatically generate bindings for known types

**Comment by latonz** [comment]
As stated by the [MS docs](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.serialization.knowntypeattribute)
> Specifies types that should be recognized by the [DataContractSerializer](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer?view=net-8.0) when serializing or deserializing a given type.
this is used by `DataContractSerializer`. Mapperly should not be coupled to the DataContractSerializer.

Could you provide an example of a mapper declaration including relevant classes and what you expect the generated code to look like?

**Comment by onionhammer** [comment]
@latonz KnownType is also used by Swagger to find derived types.

Usage could look like this
```C#

[KnownType(typeof(Dog))]
[KnownType(typeof(Cat))]
public abstract class AnimalBase
{
    public required string Name { get; set; }
}

public class Dog : AnimalBase
{
    public required string DogName { get; set; }
}

public class Cat : AnimalBase
{
    public int CatAge { get; set; }
}

public abstract class AnimalBaseDTO
{
    public string? Name { get; set; }
}

public static partial class Mapper
{
    public static partial List<AnimalBaseDTO>? MapAnimal(List<AnimalBase>? input);
}
```

Mapperly's automatic implementation could look like this

```C#
public static partial class ModelMapper
{
    public static partial List<AnimalBaseDTO>? MapAnimal(List<AnimalBase>? input)
    {
        if (input == null)
            return null;

        return new List<AnimalBaseDTO>(
            input.Select(p => p switch
            {
                Dog dog => MapDog(dog),
                Cat cat => MapCat(cat),
                _ => throw new NotSupportedException(),
            })
        );
    }

    public static MapDog(Dog input)
    {
        return new AnimalBaseDTO
        {
            Name = input.Name,
            DogName = input.DogName,
        };
    }

    public static MapCat(Cat input)
    {
        return new AnimalBaseDTO
        {
            Name = input.Name,
            CatAge = input.CatAge,
        };
    }
}

```

**Comment by onionhammer** [comment]
@latonz In my example yes, perhaps something like that would be needed, but I omitted 'KnownTypes' from the destination; I think if "KnownType" were included on the destination type, additional mapper attributes wouldnt be required.

It would be good to have the flexibility to do it in either place though

**Comment by onionhammer** [comment]
> But how should Mapperly then figure out which source type maps to which target type? Only by name/convention (eg. by replacing the dto suffix or similar)? This seems shaky to me and may not work in a lot of cases.

Yeah good point.. I guess I figured you were already doing that kind of thing, but I guess you're only doing that kind of mapping based on property names, not type names.

**Comment by latonz** [comment]
As source generators need to target netstandard2.0 I don't think generic attributes are possible, are they?

**Comment by onionhammer** [comment]
@latonz I haven't tried.. but I think it's more of a C# version 11 feature than a .NET 7 feature, so generic attributes should work?

![image](https://user-images.githubusercontent.com/969938/231839953-6d4f7759-b2af-4b8e-8486-74fd62879b95.png)

**Comment by latonz** [comment]
That's true, but that would require Mapperly and projects using Mapperly to use at least c# v11.

**Comment by onionhammer** [comment]
Could you can use preprocessor defs for it?

**Comment by latonz** [comment]
AFAIK a source generators cannot multitarget (they always need to target .NET Standard 2.0). I need to test whether multitargeting works with our `Riok.Mapperly.Abstractions` package (I don‘t think that works either, since it is a dependency of the source generator which targets .NET Standard 2.0).
Also conditional preprocessing on the Roslyn version may not work since the source generator probably runs in the context of the IDE with a potentionally higher Roslyn version than the target project uses. One approach that would work could be to generate the attributes out of the source generator. However, this would require some bigger changes which I think are not worth it as of now.

**Comment by onionhammer** [comment]
@latonz Yeah the more good points you bring up, the more it sounds like it should be a separate issue - sorry about that!

I do think it would be a nice DX feature in the future

**Comment by latonz** [comment]
@onionhammer no worries at all, I'm always thankful for new inputs and issues are here to discuss ideas 😊

**Comment by jnyrup** [comment]
A perhaps simpler case is if using Mapperly to generate a deep cloner.
Wouldn't `TSource` then trivially be equal to `TDest`?

(You might want to hide the comments regarding generic attributes to keep the thread manageable)

**Comment by latonz** [comment]
@jnyrup that's true. I don't think it is worth to introduce a separate attribute for this, as Mapperly's primary use case is object-to-object mapping and the verbosity in this case isn't much higher to configure Mapperly to just clone objects.
